lib/fizzy/execute.cpp

```diff
@@ -671,107 +652,66 @@ execution_result execute(
-            labels.push(label);
             break;
-        }
         case Instr::if_:
         {
-            const auto arity = read<uint8_t>(immediates);
-            const auto target_pc = read<uint32_t>(immediates);
-            const auto target_imm = read<uint32_t>(immediates);
-
             if (static_cast<uint32_t>(stack.pop()) != 0)
-            {
                 immediates += 2 * sizeof(uint32_t);  // Skip the immediates for else instruction.
-
-                LabelContext label{code.instructions.data() + target_pc,
-                    code.immediates.data() + target_imm, arity, stack.size()};
-                labels.push(label);
-            }
             else
             {
-                const auto target_else_pc = read<uint32_t>(immediates);
-                const auto target_else_imm = read<uint32_t>(immediates);
+                const auto target_pc = read<uint32_t>(immediates);
+                const auto target_imm = read<uint32_t>(immediates);
 
-                if (target_else_pc != 0)  // If else block defined.
-                {
-                    LabelContext label{code.instructions.data() + target_pc,
-                        code.immediates.data() + target_imm, arity, stack.size()};
-                    labels.push(label);
-                    pc = code.instructions.data() + target_else_pc;
-                    immediates = code.immediates.data() + target_else_imm;
-                }
-                else  // If else block not defined go to end of if.
-                {
-                    assert(arity == 0);  // if without else cannot have type signature.
-                    pc = code.instructions.data() + target_pc;
-                    immediates = code.immediates.data() + target_imm;
-                }
+                pc = code.instructions.data() + target_pc;
+                immediates = code.immediates.data() + target_imm;
             }
             break;
         }
         case Instr::else_:
         {
-            // We reach else only at the end of if block.
-            assert(!labels.empty());
-            const auto label = labels.top();
-            labels.pop();
+            // We reach else only after executing if block ("then" part),
+            // so we need to skip else block now.
+            const auto target_pc = read<uint32_t>(immediates);
+            const auto target_imm = read<uint32_t>(immediates);
 
-            pc = label.pc;
-            immediates = label.immediate;
+            pc = code.instructions.data() + target_pc;
+            immediates = code.immediates.data() + target_imm;
 
             break;
         }
         case Instr::end:
         {
-            // Labels can be already empty when we reach final `end`  via `return` or `br MAX`.
-            if (!labels.empty())
-                labels.pop();
-            if (labels.empty())
+            if (pc == &code.instructions[code.instructions.size()])
```

Comment:
Should have a comment this is about aborting after the "final" end.

Reply:
added comment